### PR TITLE
Feature/login with okta

### PR DESCRIPTION
**Pull request by jamieknight-db**
**PR Checklist**

- [ ] A description of the changes is added to the description of this PR.
- [ ] If there is a related issue, make sure it is linked to this PR.
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added or modified a feature, documentation in `docs` is updated

**Description of changes**

UI portion of login with Okta using okta signin widget. If the user is already signed in to Okta, they will be auto logged in to UC once they click the login button on the UC login page. If they are not logged in to Okta, a modal with sign in form will render once they click the login button on the UC login page.

There is a scenario where if the modal is closed and reopened, the widget form disappears after the re-render. Working on putting together a testing document. There are quite a lot of requirements within the okta organization itself for the widget to work.

### CODE REVIEW

**Comment by dennyglee** [comment]
Thanks very much @jamieknight-db!  For folks that are interested in helping, it would be great for folks who have access to Okta to test this out so we can clear out various corner cases.  Please chime in here and/or feel free to ping us on Slack directly as well.

**Comment by ognis1205** [comment]
If Okta has a free trial, I’ll look into it myself.

**Comment by ognis1205** [comment]
It seems that registering for Okta's free trial might not be possible with a personal email, so I'll try reproducing the issue using Keycloak's Docker image. Similarly, it might be a good idea to consider adding tests for the OAuth flow using Keycloak's Docker image.

File: ui/src/index.css [hunk 1]
```diff
@@ -1 +1,9 @@
 @import '~antd/dist/reset.css';
+/* copied from node modules package based on solution from this issue */
+/* https://github.com/okta/okta-signin-widget/issues/2429 */
+@import '../src/components/login/okta-css/dist/css/okta-sign-in.min.css';
```
Comment (ognis1205) [inline ui/src/index.css:4]: @jamieknight-db

This is not a standard testing method, but referencing [this thread](https://devforum.okta.com/t/mocking-okta-service/11941), I am trying to test this PR with a Keycloak Docker container. I am getting the following error:

Can't resolve '../src/components/login/okta-css/dist/css/okta-sign-in.min.css' in '/path/to/reviewing/repositories/unitycatalog/ui/src'

As mentioned https://github.com/okta/okta-signin-widget/issues/2429, did you copy, paste and include the CSS from the dist folder in this PR? Or am I possibly overlooking something?

Comment (jamieknight-db) [reply ui/src/index.css:4]: Hey @ognis1205 - yes I had to copy/paste the css file into the project. I am open to importing a different way if you find one that works!

Comment (ognis1205) [reply ui/src/index.css:4]: Hi @jamieknight-db,

The [original issue](https://github.com/okta/okta-signin-widget/issues/2429) is still open, but this bug was addressed and resolved in [this PR](https://github.com/okta/okta-signin-widget/pull/3634). Based on [this comment](https://github.com/okta/okta-signin-widget/pull/3618#issuecomment-2144405500), I was able to fix the issue by importing the CSS as follows:

```css
@import '@okta/okta-signin-widget/css/okta-sign-in.min.css';
```

Could you please update the implementation accordingly and remove the CSS copied under `src/`? I’ve also left a comment on the original issue suggesting it be closed.

**Comment by ognis1205** [comment]
@jamieknight-db cc: @dennyglee 

Until yesterday, I was planning to implement a mock for Okta service with (nginx + keycloak), but I struggled a bit because Okta's authentication flow uses its own OAuth2/OIDC extension called the Interaction Code Flow. I was finally able to sign up for Okta, so I should be able to proceed with a proper review today. I apologize for the delay in helping out.

File: ui/src/index.css (hunk 1 above)
Comment (creechy) [reply ui/src/index.css:4]: I think I'm seeing this as well.

**Review by creechy — APPROVED** [review]
With @ognis1205 's fix, this worked for me. We will need to document how to configure Okta to work with this. The Okta app/configuration I had, which works for the `cli`, doesn't work for the `ui` and needed to add a new app/configuration.

**Comment by ognis1205** [comment]
@jamieknight-db 

There's no urgency, but would it be possible to share the testing document mentioned in the PR?

Regarding the `setTimeout` implementation, I was wondering if an approach like

```typescript
  const isErrorHandledRef = useRef(false); 

  const widget = new OktaSignIn({
    ...
  });

  useEffect(() => {
    if (isErrorHandledRef.current) return;

    widget
      .showSignInToGetTokens({
        el: '#osw-container',
      })
      .then(function (res) {
        onSuccess(res?.idToken?.idToken);
        widget.remove();
      })
      .catch(function (error) {
        if (!isErrorHandledRef.current) {
          isErrorHandledRef.current = true;
          onError(error);
        }
      });

    return () => widget.remove();
  }, [widget, onSuccess, onError]);
```

—monitoring the `widget` itself instead of `widgetModalOpen`— might be more appropriate, though I’m not entirely sure. In any case, even if the `setTimeout` implementation is indeed the correct workaround, I believe it would be helpful to include comments since it is not a straightforward implementation.

**Comment by ognis1205** [comment]
@jamieknight-db 

I'll also give it a try and create a test document, including the Okta configuration.

File: ui/src/index.css (hunk 1 above)
Comment (jamieknight-db) [reply ui/src/index.css]: @ognis1205 I pushed up this change, let me know if it works for you. The import path ended up being a little different (included `/dist` folder).

Comment (ognis1205) [reply ui/src/index.css]: Hey @jamieknight-db,

If I haven't overlooked or misunderstood anything, I am still encountering the following errors:

```sh
% yarn start 

Failed to compile.

Error: Package path ./dist/css/okta-sign-in.min.css is not exported from package /Users/shingo/Review/unitycatalog/ui/node_modules/@okta/okta-signin-widget (see exports field in /Users/shingo/Review/unitycatalog/ui/node_modules/@okta/okta-signin-widget/package.json)
ERROR in ./src/index.css (./node_modules/css-loader/dist/cjs.js??ruleSet[1].rules[1].oneOf[5].use[1]!./node_modules/postcss-loader/dist/cjs.js??ruleSet[1].rules[1].oneOf[5].use[2]!./node_modules/source-map-loader/dist/cjs.js!./src/index.css)
Module build failed (from ./node_modules/css-loader/dist/cjs.js):
Error: Package path ./dist/css/okta-sign-in.min.css is not exported from package /Users/shingo/Review/unitycatalog/ui/node_modules/@okta/okta-signin-widget (see exports field in /Users/shingo/Review/unitycatalog/ui/node_modules/@okta/okta-signin-widget/package.json)
    at /Users/shingo/Review/unitycatalog/ui/node_modules/enhanced-resolve/lib/ExportsFieldPlugin.js:117:7
    at Hook.eval [as callAsync] (eval at create (/Users/shingo/Review/unitycatalog/ui/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:22:1)
    at Hook.CALL_ASYNC_DELEGATE [as _callAsync] (/Users/shingo/Review/unitycatalog/ui/node_modules/tapable/lib/Hook.js:18:14)
    at Resolver.doResolve (/Users/shingo/Review/unitycatalog/ui/node_modules/enhanced-resolve/lib/Resolver.js:715:16)
    at /Users/shingo/Review/unitycatalog/ui/node_modules/enhanced-resolve/lib/DescriptionFilePlugin.js:76:17
    at /Users/shingo/Review/unitycatalog/ui/node_modules/enhanced-resolve/lib/DescriptionFileUtils.js:148:13
    at /Users/shingo/Review/unitycatalog/ui/node_modules/enhanced-resolve/lib/forEachBail.js:39:13
    at onJson (/Users/shingo/Review/unitycatalog/ui/node_modules/enhanced-resolve/lib/DescriptionFileUtils.js:133:6)
    at Array.<anonymous> (/Users/shingo/Review/unitycatalog/ui/node_modules/enhanced-resolve/lib/DescriptionFileUtils.js:86:7)
    at runCallbacks (/Users/shingo/Review/unitycatalog/ui/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:45:15)

webpack compiled with 1 error
Files successfully emitted, waiting for typecheck results...
Issues checking in progress...
No issues found.
```

```sh
% yarn build

yarn run v1.22.22
$ react-scripts build
Creating an optimized production build...
Failed to compile.

Error: Package path ./dist/css/okta-sign-in.min.css is not exported from package /Users/shingo/Review/unitycatalog/ui/node_modules/@okta/okta-signin-widget (see exports field in /Users/shingo/Review/unitycatalog/ui/node_modules/@okta/okta-signin-widget/package.json)


error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Here are the versions of the Okta-related packages and Webpack I have installed:

```sh
% yarn list --pattern okta 
yarn list v1.22.22
├─ @okta/okta-auth-js@7.10.1
└─ @okta/okta-signin-widget@7.27.3
✨  Done in 0.14s.
% yarn list webpack
yarn list v1.22.22
warning Filtering by arguments is deprecated. Please use the pattern option instead.
└─ webpack@5.94.0
✨  Done in 0.13s.
```

Based on the `package.json` of the `okta-signin-widget` package, using `@import '@okta/okta-signin-widget/css/okta-sign-in.min.css';` appears to be the correct approach:

```json
% cat /Users/shingo/Review/unitycatalog/ui/node_modules/@okta/okta-signin-widget/package.json
{
 :
    "exports": {
        "./package.json": "./package.json",
        ".": {
            "types": "./types/src/exports/default.d.ts",
            "import": "./dist/esm/src/exports/exports/default.js",
            "require": "./dist/js/okta-sign-in.js",
            "default": "./dist/js/okta-sign-in.js"
        },
        "./classic": {
            "types": "./types/src/exports/classic.d.ts",
            "import": "./dist/esm/src/exports/exports/classic.js",
            "require": "./dist/js/okta-sign-in.classic.js",
            "default": "./dist/js/okta-sign-in.classic.js"
        },
        "./oie": {
            "types": "./types/src/exports/oie.d.ts",
            "import": "./dist/esm/src/exports/exports/oie.js",
            "require": "./dist/js/okta-sign-in.oie.js",
            "default": "./dist/js/okta-sign-in.oie.js"
        },
        "./polyfill": "./dist/js/okta-sign-in.polyfill.js",
        "./css/*": "./dist/css/*",
        "./mocks": "./playground/mocks/server.js"
    },
:
}
```

Please let me know if I missed anything or if further details are required.

**Comment by ognis1205** [comment]
Hey @jamieknight-db - I have successfully completed the Okta setup, and I am currently able to log in. However, I have not yet tested reproducing the specific bug scenario. If it is not urgent, I would appreciate the opportunity to investigate further, including exploring alternatives to the `setTimeout` implementation.

File: ui/src/components/login/OktaAuthButton.tsx [hunk 2]
```diff
@@ -11,31 +11,32 @@ export default function OktaAuthButton({
...
-    widget.renderEl(
-      //@ts-ignore
-      { el: widgetRef.current },
-      onSuccess,
-      onError,
-    );
-
-    return () => widget.remove();
-  }, [widgetModalOpen, onSuccess, onError]);
+    setTimeout(() => {
```
Comment (ognis1205) [inline ui/src/components/login/OktaAuthButton.tsx:26]: I believe I've identified the cause of this bug. The modal is actually not being re-rendered as expected. Instead, the issue lies in the fact that the Okta Sign-In Widget is causing duplicate DOM elements to be added via jQuery when the modal is opened and closed.

To avoid this, you can refactor the `OktaSignInWidget` into a separate component, as shown in the Gist below:

https://gist.github.com/ognis1205/b72664260fbc30daa3c5412e90164530

Then, in the `handleOnClick` function, update it as follows:

```tsx
const handleOnClick = () => {
  setWidgetModalOpen(true);
};
```

Additionally, set the `destroyOnClose={true}` property in the `Modal` component. This should ensure the modal re-renders as expected:

```tsx
<Modal
  open={widgetModalOpen}
  footer={null}
  destroyOnClose={true}
  onCancel={() => {
    setWidgetModalOpen(false);
  }}
>
  <OktaSignInWidget onSuccess={onSuccess} onError={onError} />
</Modal>
```

**Review by ognis1205 — COMMENTED** [review]
It seemed that the bug wasn't fixed in the current PR, so I investigated the issue related to `setTimeout` and made some suggestions based on my findings. What do you think?

File: ui/.eslintignore
Comment (ognis1205) [inline ui/.eslintignore:1]: Since eslint v8.25.0, `.eslintignore` has been deprecated, so I think it’s better to remove it.

**Comment by jamieknight-db** [comment]
@ognis1205 thanks so much for your suggestions and fixes. I am hoping to get back to this this week.

File: ui/src/index.css (hunk 1 above)
Comment (jamieknight-db) [reply ui/src/index.css]: Ok I updated the import again and pushed up the change. Hopefully this is resolved 🤞

File: ui/src/components/login/OktaAuthButton.tsx (hunk 2 above)
Comment (jamieknight-db) [reply ui/src/components/login/OktaAuthButton.tsx]: Thanks so much @ognis1205! I have pushed up your suggested changes.

Comment (ognis1205) [reply ui/src/components/login/OktaAuthButton.tsx]: @jamieknight-db 

My free trial expired, so I created a new account for testing. I'll review it within the week, so please give me a little more time 🙏

File: ui/src/components/login/OktaSignInWidget.tsx [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+      })
+      .then(function (res) {
+        onSuccess(res?.idToken?.idToken);
+        widget.remove();
+      })
+      .catch(function (error) {
+        onError(error);
+      });
+    return () => widget.remove();
+  }, [onSuccess, onError]);
```
Comment (ognis1205) [inline ui/src/components/login/OktaSignInWidget.tsx:33]: I also added onSuccess and onError to the dependency array in my Gist, so I share some responsibility. However, since we want this hook to run only once on mount, can we make the dependency array empty? With the current implementation, it ends up making multiple API calls.

```suggestion
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
```

**Review by ognis1205 — CHANGES_REQUESTED** [review]
@jamieknight-db 

I have one minor change request. Other than that, LGTM! Thanks for all the hard work on the implementation!
